### PR TITLE
example(Widgets): add a minimap widget

### PR DESCRIPTION
**Pull request by mgermerie**
## Description

Add a widget to display a minimap.

This widget is only supported on a `GlobeView`, and the displayed minimap is always a `PlanarView`.

### CODE REVIEW

File: examples/widgets_minimap.html [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                    url: 'http://wxs.ign.fr/decouverte/geoportail/wmts',
+                    name: "ORTHOIMAGERY.ORTHOPHOTOS",
+                    format: 'image/jpeg',
+                    tileMatrixSet: 'PM',
+                    crs: 'EPSG:3857',
+                    extent: new itowns.Extent(
+                        'EPSG:3857',
+                        -20026376.39, 20026376.39,
+                        -20048966.10, 20048966.10
+                    ),
```
Comment (gchoqueux) [inline examples/widgets_minimap.html]: could use by default  world extent ?
Global extent PM is in [Extent.js](https://github.com/iTowns/itowns/blob/ec56fa9d78310afa6d581b5deff0b252266e274e/src/Core/Geographic/Extent.js#L46)

File: examples/widgets_minimap.html [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+            // ---------- ADD INTERACTION WITH MINIMAP : ----------
+
+            // When double-clicking the minimap, travel to the cursor location.
+            const cursorCoordinates = new itowns.Coordinates(minimap.view.referenceCrs);
+            const cursorPosition = new itowns.THREE.Vector3();
+            minimap.domElement.addEventListener('dblclick', (event) => {
+                if (minimap.view.getPickingPositionFromDepth(minimap.view.eventToViewCoords(event), cursorPosition)) {
+                    cursorCoordinates.setFromVector3(cursorPosition);
+                    view.controls.lookAtCoordinate({ coord: cursorCoordinates });
```
Comment (gchoqueux) [inline examples/widgets_minimap.html]: could create a PR to add `pickCoordinates(mouseEvent or mousePosition)` and replace `getPickingPositionFromDepth`?

File: utils/gui/Minimap.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+     * @param   {View}                  view                                    The iTowns view the minimap should be
+                                                                                * linked to.
+     * @param   {ColorLayer}            layer                                   The {@link ColorLayer} that should be
+                                                                                * displayed on the minimap.
+     * @param   {Object}                [options]                               The minimap optional configuration.
+     * @param   {HTMLElement}           [options.parentElement=view.domElement] The parent HTML container of the div
+                                                                                * which contains minimap widgets.
+     * @param   {number}                [options.minAltitude=5E6]               The minimal altitude the minimap camera
+                                                                                * can reach.
+     * @param   {number}                [options.maxAltitude=5E6]               The minimal altitude the minimap camera
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: could replace by max zoom or min zoom?

File: utils/gui/Minimap.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        this.parentElement = options.parentElement || view.domElement;
+
+        this.position = options.position || DEFAULT_OPTIONS.position;
+        if (!['top-left', 'top-right', 'bottom-left', 'bottom-right'].includes(this.position)) {
+            console.warn(
+                '\'position\' optional parameter for \'Minimap\' constructor is not a valid option. '
+                + `It will be set to '${DEFAULT_OPTIONS.position}'.`,
+            );
+            this.position = DEFAULT_OPTIONS.position;
+        }
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: could you factorize with other widget?
may be add parent `Class Widget`?

File: utils/gui/Minimap.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        this.minAltitude = options.minAltitude === undefined ? DEFAULT_OPTIONS.minAltitude : options.minAltitude;
+        this.maxAltitude = options.maxAltitude === undefined ? DEFAULT_OPTIONS.maxAltitude : options.maxAltitude;
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: could you replace by zoom?

File: utils/gui/Minimap.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        this.domElement.classList.add(`${positionArray[1]}-widget`);
+
+        // Size widget according to options.
+        this.domElement.style.width = `${options.width || DEFAULT_OPTIONS.width}px`;
+        this.domElement.style.height = `${options.height || DEFAULT_OPTIONS.height}px`;
+
+        // Translate widget div according to optional translate parameter.
+        if (options.translate) {
+            this.domElement.style.transform = `translate(${options.translate.x || 0}px, ${options.translate.y || 0}px)`;
+        }
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: is possible to move in `Widget Class`?

File: utils/gui/Minimap.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        view.addFrameRequester(MAIN_LOOP_EVENTS.AFTER_RENDER, () => {
+            const mainCameraTargetPosition = view.controls.getCameraTargetPosition();
+
+            this.view.camera.camera3D.position.copy(
+                mainViewCoordinates.setFromVector3(
+                    mainCameraTargetPosition,
+                ).as(
+                    this.view.referenceCrs,
+                ).toVector3(),
+            );
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: could you simplify and add instances to copy intermediate objects?

File: utils/gui/Minimap.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+            this.view.camera.camera3D.position.z = Math.min(
+                Math.max(
+                    view.camera.camera3D.position.distanceTo(mainCameraTargetPosition),
+                    this.minAltitude,
+                ),
+                this.maxAltitude,
+            );
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: could you use zoom instead of altitude?

**Review by gchoqueux — COMMENTED** [review]
Could you use map instead of ortho image?
Could you increase size of mini map?

File: examples/widgets_minimap.html (hunk 2 above)
Comment (mgermerie) [reply examples/widgets_minimap.html]: Added in PR #1784.

File: examples/widgets_minimap.html (hunk 1 above)
Comment (mgermerie) [reply examples/widgets_minimap.html]: I changed the layer.

File: utils/gui/Minimap.js (hunk 3 above)
Comment (mgermerie) [reply utils/gui/Minimap.js]: I introduced a maximal and minimal pixel size instead.

File: utils/gui/Minimap.js (hunk 4 above)
Comment (mgermerie) [reply utils/gui/Minimap.js]: I created a `Widget` class and inherited it in other classes.

File: examples/widgets_minimap.html [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+
+
+            // ---------- ADD INTERACTION WITH MINIMAP : ----------
+
+            // When double-clicking the minimap, travel to the cursor location.
+            const cursorCoordinates = new itowns.Coordinates(minimap.view.referenceCrs);
+            const cursorPosition = new itowns.THREE.Vector3();
+            minimap.domElement.addEventListener('dblclick', (event) => {
+                if (minimap.view.getPickingPositionFromDepth(minimap.view.eventToViewCoords(event), cursorPosition)) {
```
Comment (gchoqueux) [inline examples/widgets_minimap.html]: could use the new method `pickCoordinates`?

File: utils/gui/Minimap.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+     * @param   {View}                  view                                    The iTowns view the minimap should be
+                                                                                * linked to.
+     * @param   {ColorLayer}            layer                                   The {@link ColorLayer} that should be
+                                                                                * displayed on the minimap.
+     * @param   {Object}                [options]                               The minimap optional configuration.
+     * @param   {HTMLElement}           [options.parentElement=view.domElement] The parent HTML container of the div
+                                                                                * which contains minimap widgets.
+     * @param   {number}                [options.minPixelSize=2000]             The minimal size in meters a pixel of
+                                                                                * the minimap can reach.
+     * @param   {number}                [options.maxPixelSize=Infinity]         The maximal size in meters a pixel of
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: I think it's better Scale instead of pixel size

**Comment by gchoqueux** [comment]
There's a issue with image resolution
image

File: examples/widgets_minimap.html (hunk 9 above)
Comment (mgermerie) [reply examples/widgets_minimap.html]: I now use this method.

File: utils/gui/Minimap.js (hunk 10 above)
Comment (mgermerie) [reply utils/gui/Minimap.js]: I replaced by Scale

File: utils/gui/Minimap.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+        // Coordinates used to transform position vectors from the main view CRS to the minimap view CRS.
+        const mainViewCoordinates = new Coordinates(view.referenceCrs);
+        const viewCoordinates = new Coordinates(this.view.referenceCrs);
+
+        view.addFrameRequester(MAIN_LOOP_EVENTS.AFTER_RENDER, () => {
+            // Update minimap camera zoom
+            camera3D.zoom = this.zoomRatio * maxZoom * view.getScaleFromDistance(
+                options.pitch, view.camera.camera3D.position.distanceTo(view.controls.getCameraTargetPosition()),
+            ) / this.minScale;
```
Comment (gchoqueux) [inline utils/gui/Minimap.js]: could you write?

```js
const targetPosition = view.controls.getCameraTargetPosition();

view.addFrameRequester(MAIN_LOOP_EVENTS.AFTER_RENDER, () => {
  const distance = camera3D.position.distanceTo(targetPosition);
  const scale = view.getScaleFromDistance(options.pitch, distance);
  camera3D.zoom = this.zoomRatio * maxZoom * scale / this.minScale;

```

Comment (mgermerie) [reply utils/gui/Minimap.js]: I changed this part accordingly.

**Review by gchoqueux — APPROVED** [review]

**Pull request merged**